File: android/app/src/main/kotlin/com/transport_system/ts_driver/pusher/manager/PusherManager.kt
package com.transport_system.ts_driver.pusher.manager

import android.content.Context
import com.pusher.client.Pusher
import com.pusher.client.PusherOptions
import com.pusher.client.connection.ConnectionState
import com.pusher.client.util.HttpChannelAuthorizer
import com.transport_system.ts_driver.data_providers.MyDetails
import com.transport_system.ts_driver.helpers.AppLogger

class PusherManager {


    private var pusher: Pusher? = null

    companion object {
        val instance = PusherManager()
    }

    fun initialize(context: Context) {
        try {
            if(isConnected()){
                return
            }
            val myDetails = MyDetails.loadFromSharedPrefs(context = context)
            pusher = Pusher(getAppKey(context = context), PusherOptions().apply {
                setHost(getHost(context = context))
                setWssPort(getPort(context = context))
                setUseTLS(true)
                setChannelAuthorizer(HttpChannelAuthorizer(getAuthUrl(context = context)).apply {
                    setHeaders(
                        mapOf(
                            "Authorization" to "Bearer ${myDetails?.token}",
                            "Accept" to "application/json"
                        )
                    )
                })
            })
            pusher?.connect(PusherConnectionStateListener(pusherManager = this))
        }
        catch (e:Exception){
            AppLogger.log("Exception while initializing the pusher ====> ${e.message}")
        }
    }

    fun getPusher(context: Context) : Pusher?{
        if (pusher == null){
            initialize(context = context)
        }
        return pusher
    }

    private fun isConnected() : Boolean {
        return pusher?.connection?.state == ConnectionState.CONNECTED
    }

    fun ensureConnection(context: Context) {
        try {
            if(!isConnected()){
                if(pusher == null){
                    initialize(context = context)
                    return
                }
                pusher?.connect()
            }
            else{
                AppLogger.log("Pusher socket is already connected.")
            }
        }
        catch (e:Exception){
            AppLogger.log("Exception while ensure pusher socket connection called ===> ${e.message} ")
        }
    }

    // Realtime config comes from the server (mirrored into SharedPreferences by
    // Dart). We honor it as the source of truth; the env-based hardcoded values
    // are only a fallback for the window before the first config fetch.
    private fun getAppKey(context: Context): String =
        MyDetails.realtimeKey(context = context) ?: when {
            MyDetails.isProduction(context = context) -> "HiXzy5MIniPeS24McIZ1VdIrZ"
            MyDetails.isStaging(context = context) -> "E17C5BA4D29185A3"
            else -> "9dbf5c6a056f4d4f99561fcf43f8a566" // development
        }

    private fun getHost(context: Context): String =
        MyDetails.realtimeHost(context = context) ?: when {
            MyDetails.isProduction(context = context) -> "socket.ts-portal.com"
            MyDetails.isStaging(context = context) -> "staging.ts-portal.com"
            else -> "dev.ts-portal.com" // development
        }

    private fun getPort(context: Context): Int =
        MyDetails.realtimePort(context = context) ?: 2096

    private fun getAuthUrl(context: Context): String =
        MyDetails.realtimeAuthUrl(context = context) ?: "https://${
            if (MyDetails.isStaging(context = context)) "staging." else if (MyDetails.isDevelopment(
                    context = context
                )
            ) "dev." else ""
        }ts-portal.com/broadcasting/auth"

}

File: ios/Runner/Pusher/PusherManager.swift
//
//  PusherManager.swift
//  Runner
//
//  Created by Hashim Khan on 24/01/2024.
//


import Foundation
import PusherSwift


class PusherManager : ObservableObject {
    
    
    static let shared = PusherManager()
    
    
    private var pusher: Pusher?
    // Realtime config comes from the server (mirrored into UserDefaults by Dart).
    // Honor it as the source of truth; env-based hardcoded values are only a
    // fallback for the window before the first config fetch.
    private var pusherKey: String {
        if let key = MyDetails.realtimeKey() { return key }
        if MyDetails.isProduction() { return "HiXzy5MIniPeS24McIZ1VdIrZ" }
        if MyDetails.isStaging() { return "E17C5BA4D29185A3" }
        return "9dbf5c6a056f4d4f99561fcf43f8a566" // development
    }

    
    // MARK: - Connection Functions
    func initializePusher() {
        
        // Create the Pusher ClientOptions with the custom URL
        let clusterOptions = PusherClientOptions(
            authMethod: AuthMethod.authRequestBuilder(authRequestBuilder: self),
            host: .host(getHost()),
            port: getPort(),
            useTLS: true
        )
        
        pusher = Pusher(key: pusherKey, options: clusterOptions)
        pusher?.connection.delegate = self
        
        // Establish connection
        pusher?.connect()
    }
    
    func disconnect() {
        pusher?.disconnect()
    }
    
    func isConnected() -> Bool {
        return pusher?.connection.connectionState == ConnectionState.connected
    }
    
    func ensureConnection() {
        if !isConnected() {
            printLog("Socket is not connected, attempting to reconnect...")
            guard let pusherInstance = pusher else {
                initializePusher()
                return
            }
            pusherInstance.connect()
        } else {
            printLog("Socket is already connected.")
        }
    }
    
    func getPusher() -> Pusher{
        guard let pusherInstance = pusher else {
            initializePusher()
            return pusher!
        }
        return pusherInstance
    }

    // MARK: - Configs Functions
    private func getHost() -> String {
        if let host = MyDetails.realtimeHost() { return host }
        if MyDetails.isProduction() { return "socket.ts-portal.com" }
        if MyDetails.isStaging() { return "staging.ts-portal.com" }
        return "dev.ts-portal.com" // development
    }

    private func getPort() -> Int {
        return MyDetails.realtimePort() ?? 2096
    }
    

    
    func printLog(_ message: String) {
        print("Pusher: \(message)")
    }
    
}



